File: 0974-subarray-sums-divisible-by-k/0974-subarray-sums-divisible-by-k.java
class Solution {
    public int moduler(int sum, int k) {
        int value = sum % k;
        
        if (value < 0) {
            return k + value;
        }
        
        return value;
    }
    
    public int subarraysDivByK(int[] nums, int k) {
        Map<Integer, Integer> frequency = new HashMap<>();
        int answer = 0;
        int sum = 0;
        frequency.put(0, 1);
        
        for (int i = 0; i < nums.length; i++) {
            sum += nums[i];
            int remain = moduler(sum, k);
            answer += frequency.getOrDefault(remain, 0);
            frequency.put(remain, frequency.getOrDefault(remain, 0) + 1);        
        }
        
        return answer;
    }
}

File: 0974-subarray-sums-divisible-by-k/0974-subarray-sums-divisible-by-k.cpp
class Solution {
public:
    int subarraysDivByK(vector<int>& nums, int k) {
        map<int,int>m;
        int sum=1e8;
        m[sum%k]=1;
        int answer=0;
        for(int i=0;i<nums.size();i++){
            sum+=nums[i];
            answer+=m[sum%k]++;
        }
        return answer;
    }
};